// dragon-mobile-ui/lib/Input/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Input = function (_Component) {
  _inherits(Input, _Component);

  function Input() {
    _classCallCheck(this, Input);

    return _possibleConstructorReturn(this, (Input.__proto__ || Object.getPrototypeOf(Input)).apply(this, arguments));
  }

  _createClass(Input, [{
    key: 'render',
    value: function render() {
      var _classnames;

      var props = this.props;

      var placeholder = props.placeholder,
          type = props.type,
          isRadius = props.isRadius,
          isDisabled = props.isDisabled,
          size = props.size,
          defaultValue = props.defaultValue,
          className = props.className,
          others = _objectWithoutProperties(props, ['placeholder', 'type', 'isRadius', 'isDisabled', 'size', 'defaultValue', 'className']);

      var disabled = 'disabled' in props || isDisabled;
      var radius = 'radius' in props || isRadius;

      var cls = (0, _classnames3.default)((_classnames = {
        'ui-input': true,
        'disabled': disabled,
        'radius': radius
      }, _defineProperty(_classnames, 'size-' + size, size), _defineProperty(_classnames, className, !!className), _classnames));

      var input = type == 'textarea' ? _react2.default.createElement(
        'textarea',
        _extends({}, others, { className: cls, placeholder: placeholder, disabled: disabled }),
        defaultValue
      ) : _react2.default.createElement('input', _extends({}, others, { type: type, className: cls, placeholder: placeholder, defaultValue: defaultValue, disabled: disabled }));

      var valueText = type == 'date' ? _react2.default.createElement(
        'div',
        { className: 'ui-input-placeholder' },
        placeholder
      ) : null;

      return _react2.default.createElement(
        'span',
        { className: cls },
        valueText,
        input
      );
    }
  }]);

  return Input;
}(_react.Component);

Input.propTypes = {
  type: _react.PropTypes.string,
  size: _react.PropTypes.oneOf(['xl', 'lg', 'sm', 'xs']),
  isRadius: _react.PropTypes.bool,
  isDisabled: _react.PropTypes.bool,
  className: _react.PropTypes.string
};

Input.defaultProps = {
  type: 'text',
  size: null,
  isRadius: false,
  isDisabled: false,
  className: null
};

exports.default = Input;
