// dragon-mobile-ui/lib/_Picker/PickerGroup.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _Picker = require('./Picker');

var _Picker2 = _interopRequireDefault(_Picker);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var PickerGroup = function (_Component) {
  _inherits(PickerGroup, _Component);

  function PickerGroup(props) {
    _classCallCheck(this, PickerGroup);

    var _this = _possibleConstructorReturn(this, (PickerGroup.__proto__ || Object.getPrototypeOf(PickerGroup)).call(this, props));

    _this.state = {
      visible: props.visible || false,
      value: props.value || props.defaultValue || []
    };
    return _this;
  }

  _createClass(PickerGroup, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      this.setState({
        visible: nextProps.visible,
        value: nextProps.value.concat()
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          dataSource = _props.dataSource,
          value = _props.value,
          format = _props.format,
          valueMember = _props.valueMember,
          placeholder = _props.placeholder,
          className = _props.className,
          title = _props.title,
          cancelText = _props.cancelText,
          okText = _props.okText,
          onMaskClick = _props.onMaskClick,
          onCancel = _props.onCancel,
          onOk = _props.onOk,
          onClick = _props.onClick,
          children = _props.children,
          others = _objectWithoutProperties(_props, ['dataSource', 'value', 'format', 'valueMember', 'placeholder', 'className', 'title', 'cancelText', 'okText', 'onMaskClick', 'onCancel', 'onOk', 'onClick', 'children']);

      var pickers = this.getOptions(dataSource, 0);

      var classes = (0, _classnames3.default)(_defineProperty({
        'ui-picker-container': true,
        'ui-picker-hidden': !this.state.visible
      }, className, !!className));

      var inputCls = (0, _classnames3.default)({
        'ui-picker-placeholder': !value.join(format)
      });

      return _react2.default.createElement(
        'div',
        { className: 'ui-picker-group', onClick: function onClick() {
            return _this2.toggle();
          } },
        _react2.default.createElement(
          'div',
          { className: inputCls },
          value.join(format) || placeholder
        ),
        _react2.default.createElement(
          'div',
          _extends({}, others, { className: classes, onClick: function onClick(e) {
              return _this2.onContainerClick(e);
            } }),
          _react2.default.createElement('div', { className: 'ui-picker-mask', onClick: onMaskClick }),
          _react2.default.createElement(
            'div',
            { className: 'ui-picker-inner' },
            _react2.default.createElement(
              'div',
              { className: 'ui-picker-header' },
              _react2.default.createElement(
                'div',
                { className: 'ui-picker-cancel', onClick: function onClick() {
                    return _this2.onCancel();
                  } },
                cancelText
              ),
              _react2.default.createElement(
                'div',
                { className: 'ui-picker-title' },
                title
              ),
              _react2.default.createElement(
                'div',
                { className: 'ui-picker-submit', onClick: function onClick() {
                    return _this2.onOk();
                  } },
                okText
              )
            ),
            _react2.default.createElement(
              'div',
              { className: 'ui-picker-mask-top' },
              _react2.default.createElement(
                'div',
                { className: 'ui-picker-mask-bottom' },
                _react2.default.createElement(
                  'div',
                  { className: 'ui-picker-body' },
                  _react2.default.createElement('div', { className: 'ui-picker-selected' }),
                  pickers
                )
              )
            )
          )
        )
      );
    }

    // 阻止选择器区域的默认事件

  }, {
    key: 'onContainerClick',
    value: function onContainerClick(e) {
      e.stopPropagation();
    }

    // 切换显示状态

  }, {
    key: 'toggle',
    value: function toggle() {
      this.setState({
        visible: !this.state.visible
      });
    }

    // 获取选择器组

  }, {
    key: 'getOptions',
    value: function getOptions(dataSource, level) {
      var _this3 = this;

      var _props2 = this.props,
          valueMember = _props2.valueMember,
          displayMember = _props2.displayMember;


      var pickers = this.pickers || [];
      var selected = dataSource.filter(function (item) {
        return item[valueMember] === _this3.state.value[level];
      })[0] || dataSource[0] || {};

      if (selected.children && selected.children.length > 0) {
        pickers = this.getOptions(selected.children, level + 1);
      }

      pickers.unshift(_react2.default.createElement(_Picker2.default, { key: level, valueMember: valueMember, displayMember: displayMember, dataSource: dataSource, value: selected[valueMember], onChange: function onChange(value) {
          _this3.onpickerChange(dataSource, level, value);
        } }));

      // console.log('pickers', pickers, ' pickers.length', pickers.length);
      return pickers;
    }

    // 选择器选值

  }, {
    key: 'onpickerChange',
    value: function onpickerChange(dataSource, level, value) {
      var valueMember = this.props.valueMember;


      var values = this.state.value.concat();
      var item = null;

      console.log('dataSource: ', dataSource, ' level: ', level, ' value: ', value);

      for (var i = level; i < values.length; i++) {
        item = dataSource.filter(function (item) {
          return item[valueMember] === value;
        })[0];

        console.log('item: ', item);
        values[i] = item && item[valueMember];

        dataSource = item ? item.children : [];
        value = dataSource[0] ? dataSource[0][valueMember] : undefined;
      }

      this.setState({
        value: values
      });
    }

    // getSelected(d, val) {
    //   let children = d.filter(item => item[this.props.valueMember] == val)[0].children;
    //   return children && children[0]
    // }

    // 取消

  }, {
    key: 'onCancel',
    value: function onCancel() {
      var onCancel = this.props.onCancel;

      onCancel && onCancel();
    }

    // 确定

  }, {
    key: 'onOk',
    value: function onOk() {
      var onOk = this.props.onOk;

      var value = this.state.value.concat();

      this.setState({
        value: value
      });
      onOk && onOk(value);
    }
  }]);

  return PickerGroup;
}(_react.Component);

PickerGroup.propTypes = {
  visible: _react.PropTypes.bool,
  title: _react.PropTypes.string,
  cancelText: _react.PropTypes.string,
  okText: _react.PropTypes.string,
  onMaskClick: _react.PropTypes.func,
  valueMember: _react.PropTypes.string,
  displayMember: _react.PropTypes.string
};

PickerGroup.defaultProps = {
  visible: false,
  cancelText: '取消',
  okText: '确定',
  onMaskClick: function onMaskClick() {},
  valueMember: 'value',
  displayMember: 'label',
  cascade: true
};

exports.default = PickerGroup;
