// dragon-mobile-ui/lib/Switch/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Switch = function (_Component) {
  _inherits(Switch, _Component);

  function Switch(props) {
    _classCallCheck(this, Switch);

    var _this = _possibleConstructorReturn(this, (Switch.__proto__ || Object.getPrototypeOf(Switch)).call(this, props));

    _this.state = {
      value: props.value || props.defaultValue
    };
    return _this;
  }

  _createClass(Switch, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if ('value' in nextProps) {
        this.setState({
          value: !!nextProps.value
        });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var props = this.props;

      var size = props.size,
          isCheckedText = props.isCheckedText,
          unCheckedText = props.unCheckedText,
          isDisabled = props.isDisabled,
          others = _objectWithoutProperties(props, ['size', 'isCheckedText', 'unCheckedText', 'isDisabled']);

      var value = this.state.value;

      var disabled = 'disabled' in props || isDisabled;

      var cls = (0, _classnames3.default)(_defineProperty({
        'ui-switch': true,
        'checked': value,
        'disabled': disabled
      }, 'size-' + size, !!size));

      return _react2.default.createElement(
        'span',
        _extends({ className: cls, onClick: function onClick() {
            return !disabled && _this2._onClick();
          } }, others),
        _react2.default.createElement(
          'span',
          { className: 'ui-switch-inner' },
          value ? isCheckedText : unCheckedText
        )
      );
    }
  }, {
    key: '_onClick',
    value: function _onClick() {
      var value = !this.state.value;
      this.setState({
        value: value
      });
      this.props.onChange(value);
    }
  }]);

  return Switch;
}(_react.Component);

Switch.propTypes = {
  size: _react.PropTypes.oneOf(['sm']),
  value: _react.PropTypes.bool,
  defaultValue: _react.PropTypes.bool,
  isCheckedText: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.element]),
  unCheckedText: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.element]),
  onChange: _react.PropTypes.func
};

Switch.defaultProps = {
  size: null,
  defaultValue: false,
  isCheckedText: '',
  unCheckedText: '',
  onChange: function onChange() {}
};

exports.default = Switch;
